### ISSUE TITLE
Inserting an entity with a manually specified primary key

### ISSUE THREAD

**Issue by RodH257** [ericdc1/Dapper.SimpleCRUD#39]
I have a table with an Id column that is manually specified by my code, rather than being marked as an identity column in SQL Server. Is there any way I can tell SimpleCRUD to not exclude it from the insert? The code appears to check that I have a Key or Id attribute so simply renaming the column won't work.

**Comment by clabnet** [comment]
Solved. See https://github.com/clabnet/Dapper.SimpleCRUD.
Please testing it, for me it's run, and advice me for any problems.

**Comment by ericdc1** [comment]
I reviewed this and understand the problem. I tried a few ways to deal with this by checking if the value of the Key was not 0 but it turns ugly quickly with int, long, and nul values. Clabnet's solution of using the [Required] attribute to key columns without autogenerated identity seems like a workable solution. What do you think @RodH257

**Comment by RodH257** [comment]
Sounds good to me :+1: thanks for looking into it!

**Comment by clabnet** [comment]
I will be happy if my proposal will be accepted and integrated into main development trunk.

**Comment by ericdc1** [comment]
This is in version 1.9. It was a little more complex than the @clabnet suggestion to support the various database flavors. I added unit tests to verify that this didn't break existing code. 

Thanks @clabnet for the [Required] attribute suggestion - that really helped.

**Comment by RodH257** [comment]
Thanks @ericdc1 and @clabnet for this! Looking forward to trying it out!

**Comment by clabnet** [comment]
Hi @ericdc1, 
I tried your version with my suggested implementation. It run fine but ONLY if the id autoincrement is an integer type. Yes, I known that the integer contain 2.000.000.000+ rows, but this is not known who created the database I'm working. I can't change the type id on table definitions, I must use the long type.
I would mark as deprecated my version https://github.com/clabnet/Dapper.SimpleCRUD of your code, but I can't do it until this feature is not running with this requirements.
Please feel free to contact me for other informations.
Grazie for your work.

**Comment by walterwright** [comment]
I'm having trouble making this aspect of SimpleCRUD 1.9.1 work with SQL Server 2014. My 'Id' field is a Primary key but NOT an IDENTITY field. When I set the [Required] attribute on the Id field and attempt to insert a value, an InvalidOperation exception is thrown: 'No column was selected'. I'm not able to set things up so I can trace what's happening here, but looking at the source of SimpleCRUD.cs, when used with SQL Server I believe it is still going to attach the "_getIdentitySQL' clause, correct? (because the Primary Key is an int). Does that still work if there is no IDENTITY field? I tried making Id an IDENTITY field but then I predictably get the error saying you can't insert that field if IDENTITY_INSERT is OFF. Maybe I've got the wrong end of how this feature should work, but any ideas as to what I should try would be welcome.

Here's the top part of the stack trace:
[InvalidOperationException: No columns were selected]
   Dapper.SqlMapper.GetDapperRowDeserializer(IDataRecord reader, Int32 startBound, Int32 length, Boolean returnNullIfFirstMissing) in D:\Dev\dapper-dot-net\Dapper NET40\SqlMapper.cs:2590
   Dapper.SqlMapper.GetDeserializer(Type type, IDataReader reader, Int32 startBound, Int32 length, Boolean returnNullIfFirstMissing) in D:\Dev\dapper-dot-net\Dapper NET40\SqlMapper.cs:2207
   Dapper.<QueryAsync>d__147`1.MoveNext() in D:\Dev\dapper-dot-net\Dapper NET45\SqlMapperAsync.cs:101
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +99
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +58
   Dapper.<QueryAsync>d__141.MoveNext() in D:\Dev\dapper-dot-net\Dapper NET45\SqlMapperAsync.cs:37
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +99
   System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) +58
   Dapper.<InsertAsync>d__61`1.MoveNext() in c:\Users\Eric\Documents\GitHub\Dapper.SimpleCRUD\Dapper.SimpleCRUD NET45\SimpleCRUDAsync.cs:281

**Comment by walterwright** [comment]
Further to the above, I have managed to trace through the code and it looks like the _getIdentitySQL clause is NOT being appended to the query due to line 278 in SimpleCRUDAsync.cs:
```
            if ((keytype == typeof(int) || keytype == typeof(long)) && Convert.ToInt64(idProps.First().GetValue(entityToInsert, null)) == 0)
            {
                sb.Append(";" + _getIdentitySql);
            }
```
This causes SQL Server not to return any rows and hence the Deserializer in SqlMapperAsync exceptions. I altered the code locally so that the _getIdentitySQL clause WAS appended and it worked fine, however I don't know how that would play out for the other Db Providers.

**Comment by ericdc1** [comment]
I'll check this out today. In the mean time take a look at this unit test that should cover this scenario. https://github.com/ericdc1/Dapper.SimpleCRUD/blob/master/Dapper.SimpleCRUDTests/Tests.cs#L784

It's possible the insert in the async method isn't the same as the standard insert.

**Comment by walterwright** [comment]
That test does indeed fail if I call InsertAsync instead of Insert.

**Comment by ericdc1** [comment]
Fixed with latest commit - 1.9.2 is up on Nuget.

**Comment by walterwright** [comment]
That has fixed the problem perfectly. Many thanks for sorting it out so quickly Eric.
